### ISSUE TITLE
Option to recognize TSX in files with .ts extension

### ISSUE THREAD

**Issue by Wedvich** [typescript-eslint/typescript-eslint#574]
**What code were you trying to parse?**

```jsx
// This file is named Thing.ts

const Thing = () => (
    <div>Thing</div>
)
```

**What did you expect to happen?**
It should be parsed as TSX.

**What actually happened?**
Parsing fails with `Parsing error: Unterminated regular expression literal.`

I consistently use `.ts` as the file extension both for pure TypeScript files and my React components which contain TSX. I want my files to be parsed as TSX regardless of the file extension.

The [documentation](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#configuration) specifies:

> ecmaFeatures.jsx - default false. Enable parsing JSX when true. More details can be found here.
> 
> - It's false on *.ts files regardless of this option.
> - It's true on *.tsx files regardless of this option.
> - Otherwise, it respects this option.

This unfortunately leads to the confusing situation where my code successfully transpiles, but ESLint reports an error. Is it possible to override this behaviour somehow, or could an option be added for it?

For `@babel/preset-typescript` for example, it has the same default behaviour, but by setting the options `allExtensions: true` and `isTSX: true` it can be overridden to parse everything as TSX regardless of the extension (https://babeljs.io/docs/en/babel-preset-typescript#options)

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `1.9.0` |
| `TypeScript`                | `3.4.5` |
| `ESLint`                    | `5.16.0` |
| `node`                      | `12.3.1` |
| `npm`                       | `6.9.0` |

**Comment by kaicataldo** [comment]
Out of curiosity, how are you getting around the parsing ambiguities described by https://github.com/Microsoft/TypeScript/issues/26489#issuecomment-413595484 and https://github.com/microsoft/TypeScript/issues/30503#issuecomment-475417559?

From the [official docs](https://www.typescriptlang.org/docs/handbook/jsx.html):
> Since TypeScript also uses angle brackets for type assertions, combining it with JSX’s syntax would introduce certain parsing difficulties. As a result, TypeScript disallows angle bracket type assertions in .tsx files.

**Comment by Wedvich** [comment]
Personally I stick to the `as` keyword for exactly that reason:
```ts
(instance as Thing).doStuff()
```

**Comment by bradzacher** [comment]
I don't quite understand how you're able to use JSX without the `.tsx` extension?
I thought that typescript built so that it didn't work that way.

In my usage in the past, using JSX within a `.ts` file threw a bunch of errors.

**Comment by Wedvich** [comment]
You're right, I use Babel with [@babel/plugin-transform-typescript](https://babeljs.io/docs/en/babel-plugin-transform-typescript), so it's transpiled by Babel and _not_ the official TypeScript compiler. So then I suppose the `.tsx` extension is imposed by the TypeScript compiler to switch modes internally? Is there any flag that could be used to override this when passing files from the parser to the TSC, or could the TSC be fooled to think it receives a `.tsx` file for example?

The main reason I would prefer to use only the `.ts` extension is ergonomics - before switching to TypeScript, it was a pain to have to rename files between `.js`/`.jsx` when changing their contents (especially having to do `git mv` to preserve file history) - so I stuck to `.js` across the board. I would very much prefer to keep that convention for TypeScript as well, but if it's not possible for technical reasons, I guess using `.tsx` for everything is a fallback option.

**Comment by bradzacher** [comment]
> especially having to do `git mv` to preserve file history

I was under the impression that git should be able handle file renames automatically as long as you don't change too much of the file content at the same time.

---

I assume you don't use an IDE with TS integration (like vscode) either? I haven't tried it, but I'm pretty sure that IDE integration is based of the TS language server, so it'd throw errors for your code if you don't use `.tsx`?

> So then I suppose the .tsx extension is imposed by the TypeScript compiler to switch modes internally?

AFAIK that's correct - TypeScript does use extensions to switch settings internally, afaik there's no way to change this without passing in a false filename, though for a concrete answer I'd have to defer to someone like @uniqueiniquity on that.

I think we could technically make it work (I think we pass in a false filename as it is), but I don't know enough about the internals of ts-estree to be sure (deferring to ben or @JamesHenry to answer that).

**Comment by uniqueiniquity** [comment]
That's correct; you _could_ do this by passing in a false filename, but as far as the typescript-eslint parser goes, in order to support the semantic rule scenario, we'd have to start doing work to "lie" about _every_ .ts file. I also second the point about the IDE case, as well as adding the fact that babel isn't actually typechecking your code (I expect you knew this, but I wanted to make sure it was clear).

**Comment by Wedvich** [comment]
I use VS Code and you're right, the built-in TypeScript validator complains if I don't use the `.tsx` extension. But I've set `"typescript.validate.enable": false` in my workspace to avoid duplicate validations, so it would be nice if a configuration was available for the ESLint validation.

Still, as mentioned there's no typechecking from Babel, so actually running the typechecks via tsc would still need to be solved somehow regardless of ESLint. Either way I think my use case is very limited and there wouldn't be much value in adding this before the TypeScript compiler itself has it as a configurable switch. Closing this - thank you for valuable input!

> I was under the impression that git should be able handle file renames automatically as long as you don't change too much of the file content at the same time.

You're right, it does. I don't do this for my personal projects either so I'm not really sure why I included this 🤔 We had it as a routine at a previous client project, but now I think that was more due to FUD about how Git works after switching over from TFVC.
